server/src/main/java/org/opensearch/cluster/coordination/JoinHelper.java

```diff
@@ -407,12 +459,38 @@ public String executor() {
     }
 
     public void sendValidateJoinRequest(DiscoveryNode node, ClusterState state, ActionListener<TransportResponse.Empty> listener) {
-        transportService.sendRequest(
-            node,
-            VALIDATE_JOIN_ACTION_NAME,
-            new ValidateJoinRequest(state),
-            new ActionListenerResponseHandler<>(listener, i -> Empty.INSTANCE, ThreadPool.Names.GENERIC)
-        );
+        if (node.getVersion().before(Version.V_3_0_0)) {
+            transportService.sendRequest(
+                node,
+                VALIDATE_JOIN_ACTION_NAME,
+                new ValidateJoinRequest(state),
+                new ActionListenerResponseHandler<>(listener, i -> Empty.INSTANCE, ThreadPool.Names.GENERIC)
+            );
+        } else {
+            try {
+                if (!serializedStates.containsKey(state.version())) {
+                    serializedStates.clear();
+                }
+                BytesReference bytes = serializedStates.computeIfAbsent(state.version(), version -> {
+                    try {
+                        return ClusterStateUtils.serializeClusterState(state, node, true);
+                    } catch (IOException e) {
+                        // mandatory as ConcurrentHashMap doesn't rethrow IOException.
+                        throw new RuntimeException(e);
+                    }
+                });
```

Comment:
Given that the map will only ever hold a single entry, what do you think about using an AtomicReference instead? Something like this:

```java
final AtomicReference<Tuple<Long, BytesReference>> serializedState = new AtomicReference<>();

...

final BytesReference bytes = serializedState.updateAndGet(tuple -> {
    if (tuple.v1() == state.version) {
        return tuple;
    } else {
        try {
            return new Tuple(state.version(), ClusterStateUtils.serializeClusterState(state, node, true));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
});
```

I think that works the same as the logic here. Not a huge deal, but it makes the intent a little more clear that you're only ever memoizing a single version of the cluster state.